src/MergeContext.js

```diff
@@ -856,10 +513,302 @@ class MergeContext {
+        this._log("checkApproval: " + this._approval);
+        await this._setApprovalStatus(this._prHeadSha());
+
+        if (this._prInProgress()) {
+            this._logFailedCondition("not in progress");
+            return StepResult.Fail();
+        }
+
+        if (!this._prMergeable()) {
+            this._logFailedCondition("mergeable");
+            return StepResult.Fail();
+        }
+
+        if (await this._stagingFailed()) {
+            this._logFailedCondition("fresh tag with failed staging checks");
+            return StepResult.Fail();
+        }
+
+        const statusChecks = await this._getPrStatuses();
+        if (statusChecks.failed())
+            return StepResult.Fail();
+
+        if (!messageValid) {
+            this._logFailedCondition("commit message");
+            return StepResult.Fail();
+        }
+
+        if (!this._approval.granted()) {
+            this._logFailedCondition("approved");
+            return StepResult.Fail();
+        }
+
+        if (this._approval.grantedTimeout())
+            return StepResult.Delay(this._approval.delayMs);
+
+        if (!statusChecks.final())
+            return StepResult.Suspend();
+
+        return StepResult.Succeed();
+    }
+
+    async _createStaged() {
+        this._log("start merging...");
+        const baseSha = await GH.getReference(this._prBaseBranchPath());
+        const mergeSha = await GH.getReference("pull/" + this._number() + "/merge");
+        const mergeCommit = await GH.getCommit(mergeSha);
+        if (!Config.githubUserName())
+            await this._acquireUserProperties();
+        let now = new Date();
+        const committer = {name: Config.githubUserName(), email: Config.githubUserEmail(), date: now.toISOString()};
+        const tempCommitSha = await GH.createCommit(mergeCommit.tree.sha, this._prMessage(), [baseSha], mergeCommit.author, committer);
+        this._tagSha = await GH.createReference(tempCommitSha, "refs/" + this._stagingTag());
+        await this._setApprovalStatus(this._tagSha);
+        await GH.updateReference(Config.stagingBranchPath(), this._tagSha, true);
+    }
+
+    // returns filled StepResult object
+    async process() {
+        // TODO: Optimize old/busy repo by quitting unless _prOpen().
+        // TODO: Optimize label tests by caching all PR labels here.
+
+        if (!this._dryRun("reset labels before precondition checking"))
+            await this._unlabelPreconditionsChecking();
+
+        const conditions = await this._checkConditions();
+        if (!conditions.succeeded())
+            return conditions;
+
+        if (this._dryRun("start merging"))
+            return StepResult.Suspend();
+
+        await this._unlabelPreconditionsChecked();
+        await this._createStaged();
+        await this._labelWaitingStagingChecks();
+
+        return StepResult.Succeed();
+    }
+} // MergeInitiator
+
+// Finishes PR processing.
+// Waits for staging checks completing, merges staged commit
```

Comment:
I would not try to enumerate the steps here -- the list will quickly become stale. I would generalize the second sentence into something like "Performs all PR merge steps after the staged commit creation".

Reply:
Done.